### PR TITLE
fix(gateway): don't address resources via inaccessible IPs

### PR DESCRIPTION
**Pull request by thomaseizinger**
Gateways may not run with a dual network stack. A common example here is docker where IPv6 by default does not work. Similarly, AWS does not support IPv6 very well.

Using the same technique as on the client in #6383, we can have the gateway detect its supported IP stack. If we don't have an IPv4 or IPv6 socket, we should not create entries in our translation table for those addresses. Instead, we need to cycle through the provided proxy IPs and utilise our NAT64 / NAT46 implementation to correctly route the traffic.

refs: #6371.

### CODE REVIEW

**Comment by conectado** [comment]
We never find out if there is a new nic in the gateway that has the desired ip stack

**Comment by thomaseizinger** [comment]
> We never find out if there is a new nic in the gateway that has the desired ip stack

Roaming is also not handled today because we never rebind our sockets so that isn't really a regression, is it?

Plus, this state is per connection so by reconnecting to the gateway, a user can reset it.

**Comment by conectado** [comment]
> > We never find out if there is a new nic in the gateway that has the desired ip stack
> 
> Roaming is also not handled today because we never rebind our sockets so that isn't really a regression, is it?
> 
> Plus, this state is per connection so by reconnecting to the gateway, a user can reset it.

Isn't the stack set globally for the socket?

I'm not talking about just roaming, I mean just adding a new adapter to the gateway. Right now is that happens no matter the ip stack it will be used for communicating with the external world.

**Comment by thomaseizinger** [comment]
> > > We never find out if there is a new nic in the gateway that has the desired ip stack
> > 
> > 
> > Roaming is also not handled today because we never rebind our sockets so that isn't really a regression, is it?
> > Plus, this state is per connection so by reconnecting to the gateway, a user can reset it.
> 
> Isn't the stack set globally for the socket?

The socket state yes. The filtered IPs no.

> I'm not talking about just roaming, I mean just adding a new adapter to the gateway. Right now is that happens no matter the ip stack it will be used for communicating with the external world.

Right and on Linux binding to the unspecified IP will also use that new interface.

@jamilbk Is it acceptable for now to require a restart of the gateway if the network configuration changes?

**Comment by thomaseizinger** [comment]
We could perhaps query network interfaces periodically using `networkctl`? https://man.archlinux.org/man/networkctl.1.en

**Comment by conectado** [comment]
> We could perhaps query network interfaces periodically using `networkctl`? https://man.archlinux.org/man/networkctl.1.en

Perhaps we can probe at some interval with a packet to some known ip?

Personally, I'm also fine requiring a gateway restart for any network changes.

**Comment by jamilbk** [comment]
> > > > We never find out if there is a new nic in the gateway that has the desired ip stack
> > > 
> > > 
> > > Roaming is also not handled today because we never rebind our sockets so that isn't really a regression, is it?
> > > Plus, this state is per connection so by reconnecting to the gateway, a user can reset it.
> > 
> > 
> > Isn't the stack set globally for the socket?
> 
> The socket state yes. The filtered IPs no.
> 
> > I'm not talking about just roaming, I mean just adding a new adapter to the gateway. Right now is that happens no matter the ip stack it will be used for communicating with the external world.
> 
> Right and on Linux binding to the unspecified IP will also use that new interface.
> 
> @jamilbk Is it acceptable for now to require a restart of the gateway if the network configuration changes?

Hm I can't say for certain. The safe thing to do would have it auto-heal, but I realize that is tricky to do on Linux. We would probably just use our own IPv4 and IPv6 connectivity check right? We could simply `HEAD API_URL` over v4 and v6 as a connectivity check?

The only issue with requiring a restart for network changes is that admins may not be able to easily reach them. On a discovery call earlier this week the customer specifically asked if the Gateways are managed remotely or from the CLI. Think restarting the service on 45 Gateways in case you add a new IPv6 Gateway to your cloud account or something.

**Comment by thomaseizinger** [comment]
> The only issue with requiring a restart for network changes is that admins may not be able to easily reach them. On a discovery call earlier this week the customer specifically asked if the Gateways are managed remotely or from the CLI. Think restarting the service on 45 Gateways in case you add a new IPv6 Gateway to your cloud account or something.

We already don't support roaming for gateways so this isn't really a regression IMO. I opened https://github.com/firezone/firezone/issues/6402 to track the actual implementation of this.

**Review by jamilbk — APPROVED** [review]
LGTM. I updated the desc to ref instead of fix #6371 since we might need a connectivity check to actually see if we can route IPv6 to public hosts.

This is a conservative approach though since if we can't bind then there is no possibility of routing IPv6.

**Review by conectado — APPROVED** [review]
Code LGTM if we are fine with restarting(for now) to reach new adapters that goes to external resources.

**Comment by thomaseizinger** [comment]
This builds on a few other things that I am no longer sure I want to merge. Putting this on hold.

**Comment by thomaseizinger** [comment]
Closing this as outdated. Instead, we should implement proper NAT-ing of ICMP error messages.

**Pull request closed without merging**
